### PR TITLE
Scp3114::Dancing event

### PR DESCRIPTION
**Pull request by VALERA771**

### CODE REVIEW

**Comment by VALERA771** [comment]
I think `NewState` setter can be removed. Or `NewState` at all can be

File: Exiled.Events/Patches/Events/Scp3114/Dancing.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                    // Handlers.Scp3114.OnDancing(ev);
+                    new(OpCodes.Call, Method(typeof(Handlers.Scp3114), nameof(Handlers.Scp3114.OnDancing))),
+
+                    // if (ev.IsAllowed)
+                    //   goto continueLabel;
+                    new(OpCodes.Callvirt, PropertyGetter(typeof(DancingEventArgs), nameof(DancingEventArgs.IsAllowed))),
+                    new(OpCodes.Brtrue_S, continueLabel),
+
+                    // return
+                    new(OpCodes.Ret),
```
Comment (louis1706) [inline Exiled.Events/Patches/Events/Scp3114/Dancing.cs:66]: Not very important but you can make this smaller by making the label going to ret and remove ret

Comment (VALERA771) [reply Exiled.Events/Patches/Events/Scp3114/Dancing.cs:66]: Yeah but I'm still thinking about `NewState` setter. If it'll be removed, idk if there'll be a difference

File: Exiled.Events/Patches/Events/Scp3114/Dancing.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+            Label continueLabel = generator.DefineLabel();
+
+            LocalBuilder ev = generator.DeclareLocal(typeof(DancingEventArgs));
+
+            newInstructions.InsertRange(
+                index,
+                new[]
+                {
+                    // reader.ReadBool();
```
Comment (louis1706) [inline Exiled.Events/Patches/Events/Scp3114/Dancing.cs:42]: No ?

**Review by louis1706 — CHANGES_REQUESTED** [review]

File: Exiled.Events/Patches/Events/Scp3114/Dancing.cs (hunk 2 above)
Comment (VALERA771) [reply Exiled.Events/Patches/Events/Scp3114/Dancing.cs:42]: Just showing that first constructor argument is above (inserting right after that)

Comment (louis1706) [reply Exiled.Events/Patches/Events/Scp3114/Dancing.cs:42]: > Just showing that first constructor argument is above (inserting right after that)

That should be say with Index so not inside the AddRange

File: Exiled.Events/Patches/Events/Scp3114/Dancing.cs (hunk 1 above)
Comment (louis1706) [reply Exiled.Events/Patches/Events/Scp3114/Dancing.cs:66]: > Yeah but I'm still thinking about `NewState` setter. If it'll be removed, idk if there'll be a difference

Setter for newState is not needed only keep IsAllowed

**Review by NaoUnderscore — APPROVED** [review]

**Review by louis1706 — COMMENTED** [review]
[SUGGESTION] is it possible to choose wish dance is made and possibly having an Enum for it ?

**Comment by VALERA771** [comment]
> [SUGGESTION] is it possible to choose wish dance is made and possibly having an Enum for it ?

As far as I tested, it'll cause some wierd bugs

**Comment by louis1706** [comment]
> > [SUGGESTION] is it possible to choose wish dance is made and possibly having an Enum for it ?
> 
> As far as I tested, it'll cause some wierd bugs

So you can add the Dance play as getter

**Comment by VALERA771** [comment]
> > > [SUGGESTION] is it possible to choose wish dance is made and possibly having an Enum for it ?
> > 
> > As far as I tested, it'll cause some wierd bugs
> 
> So you can add the Dance play as getter 

Not possible. Patching that method in any way will cause some bugs

File: Exiled.Events/EventArgs/Scp3114/DancingEventArgs.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        /// </summary>
+        /// <param name="newState"><inheritdoc cref="DanceType"/></param>
+        /// <param name="player"><inheritdoc cref="Player"/></param>
+        /// <param name="isAllowed"><inheritdoc cref="IsAllowed"/></param>
+        public DancingEventArgs(bool newState, Player player, bool isAllowed = true)
+        {
+            Player = player;
+            Scp3114 = player.Role.As<Scp3114Role>();
+            IsDancing = newState;
+            DanceType = (DanceType)(newState ? UnityEngine.Random.Range(0, Scp3114.Dance._danceVariants) : 0);
```
Comment (VALERA771) [inline Exiled.Events/EventArgs/Scp3114/DancingEventArgs.cs:32]: Maybe add None in enum for this case? I think there can be an error cuz enum doesn't have such value (or it has, but its Dance1)

Comment (louis1706) [reply Exiled.Events/EventArgs/Scp3114/DancingEventArgs.cs:32]: I would like to give name to the different dance but yeah 0 will be None

Comment (VALERA771) [reply Exiled.Events/EventArgs/Scp3114/DancingEventArgs.cs:32]: I also don't know to name dances. I don't know them. maybe we should leave a small description about dance and name it based on it's style?

File: Exiled.API/Enums/DanceType.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+// <copyright file="DanceType.cs" company="Exiled Team">
+// Copyright (c) Exiled Team. All rights reserved.
+// Licensed under the CC BY-SA 3.0 license.
+// </copyright>
+// -----------------------------------------------------------------------
+
+namespace Exiled.API.Enums
+{
+    /// <summary>
+    /// Gets all SCP-3114 dance types.
```
Comment (NaoUnderscore) [inline Exiled.API/Enums/DanceType.cs:11]:
```suggestion
    /// All available dance variants being used by SCP-3114.
```

File: Exiled.API/Features/Roles/Scp3114Role.cs [hunk 5]
```diff
@@ -220,6 +221,11 @@ public float WarningTime
             set => Identity._warningTimeSeconds = value;
         }
 
+        /// <summary>
+        /// Gets or sets a dance type that SCP-3114 will have./>.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Roles/Scp3114Role.cs:225]:
```suggestion
        /// Gets or sets the bound dance.
```

File: Exiled.API/Features/Roles/Scp3114Role.cs [hunk 6]
```diff
@@ -249,8 +255,14 @@ public void PlaySound(Scp3114VoiceLines.VoiceLinesName voiceLine = Scp3114VoiceL
         public void StopDancing() => Dance.SendRpc(x => x.IsDancing = false);
 
         /// <summary>
-        /// Starts dancing.
+        /// Make Scp3114 Dancing.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Roles/Scp3114Role.cs:258]:
```suggestion
        /// Starts the dancing process.
```

File: Exiled.API/Features/Roles/Scp3114Role.cs [hunk 7]
```diff
@@ -249,8 +255,14 @@ public void PlaySound(Scp3114VoiceLines.VoiceLinesName voiceLine = Scp3114VoiceL
...
         /// </summary>
-        public void StartDanging() => Dance.SendRpc(x => x.IsDancing = true);
+        /// <param name="danceType">Wish <see cref="DanceType"/> than Scp3114 will do.</param>
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Roles/Scp3114Role.cs:260]:
```suggestion
        /// <param name="danceType">The requested <see cref="DanceType"/>.</param>
```

File: Exiled.Events/EventArgs/Scp3114/DancingEventArgs.cs [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            IsDancing = newState;
+            DanceType = (DanceType)(newState ? UnityEngine.Random.Range(0, Scp3114.Dance._danceVariants) : byte.MaxValue);
+            IsAllowed = isAllowed;
+        }
+
+        /// <inheritdoc/>
+        public bool IsAllowed { get; set; }
+
+        /// <summary>
+        /// Gets a value indicating whether or not SCP-3114 will dancing.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Scp3114/DancingEventArgs.cs:40]:
```suggestion
        /// Gets a value indicating whether or not SCP-3114 is dancing.
```

File: Exiled.Events/EventArgs/Scp3114/DancingEventArgs.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        /// </summary>
+        public bool IsDancing { get; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether or not SCP-3114 will dance.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Scp3114/DancingEventArgs.cs:45]:
```suggestion
        /// Gets or sets the bound dance.
```

File: Exiled.Events/EventArgs/Scp3114/DancingEventArgs.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+namespace Exiled.Events.EventArgs.Scp3114
+{
+    using Exiled.API.Enums;
+    using Exiled.API.Features;
+    using Exiled.API.Features.Roles;
+    using Exiled.Events.EventArgs.Interfaces;
+    using Exiled.Events.Patches.Events.Scp3114;
+
+    /// <summary>
+    /// Contains all information before SCP-3114 starts or stops dancing.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Scp3114/DancingEventArgs.cs:17]:
```suggestion
    /// Contains all information before SCP-3114 changes its dancing status.
```

File: Exiled.Events/Handlers/Scp3114.cs [hunk 11]
```diff
@@ -47,6 +47,11 @@ public static class Scp3114
         /// </summary>
         public static Event<VoiceLinesEventArgs> VoiceLines { get; set; } = new();
 
+        /// <summary>
+        /// Invoked before SCP-3314 starts or stops dancing.
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Scp3114.cs:51]:
```suggestion
        /// Invoked before SCP-3314 changes its dancing status.
```

File: Exiled.Events/Handlers/Scp3114.cs [hunk 12]
```diff
@@ -88,6 +93,12 @@ public static class Scp3114
         /// <param name="ev">The <see cref="VoiceLinesEventArgs" /> instance.</param>
         public static void OnVoiceLines(VoiceLinesEventArgs ev) => VoiceLines.InvokeSafely(ev);
 
+        /// <summary>
+        /// Called before SCP-3314 starts or stops dancing.
```
Comment (NaoUnderscore) [inline Exiled.Events/Handlers/Scp3114.cs:97]:
```suggestion
        /// Called before SCP-3314 changes its dancing status.
```

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]

**Comment by VALERA771** [comment]
We still have the biggest problem - how to name dance types

**Comment by louis1706** [comment]
> We still have the biggest problem - how to name dance types

https://github.com/creepycats/EmoteForAll-SL/blob/main/Types/Scp3114DanceType.cs

**Comment by Misfiy** [comment]
Change Scp3114Role::StatDancing to Scp3114Role::StartDancing

File: Exiled.API/Features/Roles/Scp3114Role.cs [hunk 13]
```diff
@@ -249,8 +255,14 @@ public void PlaySound(Scp3114VoiceLines.VoiceLinesName voiceLine = Scp3114VoiceL
         public void StopDancing() => Dance.SendRpc(x => x.IsDancing = false);
 
         /// <summary>
-        /// Starts dancing.
+        /// Starts the dancing process.
         /// </summary>
-        public void StartDanging() => Dance.SendRpc(x => x.IsDancing = true);
+        /// <param name="danceType">The requested <see cref="DanceType"/>.</param>
+        public void StatDancing(DanceType danceType) => Dance.SendRpc((x) =>
```
Comment (ALEXWARELLC) [inline Exiled.API/Features/Roles/Scp3114Role.cs:261]: As @Misfiy stated, change this typo to `StartDancing`.

**Comment by VALERA771** [comment]
> > We still have the biggest problem - how to name dance types
> 
> 
> 
> https://github.com/creepycats/EmoteForAll-SL/blob/main/Types/Scp3114DanceType.cs

Does the numeration right? We really rely on number of enum in this case

**Comment by Misfiy** [comment]
> Does the numeration right? We really rely on number of enum in this case

Looks like it should be working

**Review by Misfiy — APPROVED** [review]

**Pull request merged**